src/frequenz/client/reporting/_client.py

```diff
@@ -0,0 +1,289 @@
+    ListMicrogridComponentsDataRequest as PBListMicrogridComponentsDataRequest,
+)
+from frequenz.api.reporting.v1.reporting_pb2 import (
+    ListMicrogridComponentsDataResponse as PBListMicrogridComponentsDataResponse,
+)
+from frequenz.api.reporting.v1.reporting_pb2 import TimeFilter as PBTimeFilter
+from frequenz.api.reporting.v1.reporting_pb2_grpc import ReportingStub
+from frequenz.client.common.metric import Metric
+from google.protobuf.timestamp_pb2 import Timestamp as PBTimestamp
+
+# pylint: enable=no-name-in-module
+
+Sample = namedtuple("Sample", ["timestamp", "value"])
+"""Type for a sample of a time series."""
+
+MetricSample = namedtuple(
+    "MetricSample", ["timestamp", "microgrid_id", "component_id", "metric", "value"]
+)
+"""Type for a sample of a time series incl. metric type, microgrid and component ID"""
+
+
+@dataclass(frozen=True)
+class ComponentsDataPage:
+    """A page of microgrid components data returned by the Reporting service."""
+
+    _data_pb: PBListMicrogridComponentsDataResponse
+    """The underlying protobuf message."""
+
+    def is_empty(self) -> bool:
+        """Check if the page contains valid data.
+
+        Returns:
+            True if the page contains no valid data.
+        """
+        if not self._data_pb.microgrids:
+            return True
+        if not self._data_pb.microgrids[0].components:
+            return True
+        if not self._data_pb.microgrids[0].components[0].metric_samples:
+            return True
+        return False
+
+    def iterate_metric_samples(self) -> Generator[MetricSample, None, None]:
+        """Get generator that iterates over all values in the page.
+
+        Note: So far only `SimpleMetricSample` in the `MetricSampleVariant`
+        message is supported.
+
+
+        Yields:
+            A named tuple with the following fields:
+            * timestamp: The timestamp of the metric sample.
+            * microgrid_id: The microgrid ID.
+            * component_id: The component ID.
+            * metric: The metric name.
+            * value: The metric value.
+        """
+        data = self._data_pb
+        for mdata in data.microgrids:
+            mid = mdata.microgrid_id
+            for cdata in mdata.components:
+                cid = cdata.component_id
+                for msample in cdata.metric_samples:
+                    ts = msample.sampled_at.ToDatetime()
+                    met = Metric.from_proto(msample.metric).name
+                    value = (
+                        msample.sample.simple_metric.value
+                        if msample.sample.simple_metric
+                        else None
+                    )
+                    yield MetricSample(
+                        timestamp=ts,
+                        microgrid_id=mid,
+                        component_id=cid,
+                        metric=met,
+                        value=value,
+                    )
+
+    @property
+    def next_page_token(self) -> Any:
```

Comment:
This should return `str`, right? If you want it to return an *opaque* type, then use `object` instead, as `Any` will completely disable type checking, doing wrong stuff like `t = p.next_page_token; t.i_dont_exist()` will not be detected by `mypy`.

Reply:
I suggest also discussing this one here, and only create a separate issue if it is not trivial.

Reply:
https://github.com/frequenz-floss/frequenz-client-reporting-python/pull/28